### PR TITLE
feat(*): support Graal.js

### PR DESCRIPTION
**Pull request by tmetzke**
* adds single element `append` and `remove` methods for `DomXmlElement`
  since Graal.js does not use varargs methods when there is a collection-based
  method as well
* adds `propList` method to `JacksonJsonNode` for
  Graal.js-compliant calls with JS arrays
* adds Graal.js test dependencies
* extracts Spin core tests for reuse
* creates explicit Nashorn and Graal.js tests

### CODE REVIEW

File: core/src/test/java/org/camunda/spin/test/script/JavascriptScriptRuleNashornTest.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+package org.camunda.spin.test.script;
+
+import org.camunda.spin.test.ScriptEngine;
+import org.camunda.spin.test.ScriptRuleTest;
+
+/**
+ * @author Sebastian Menski
+ */
```
Comment (koevskinikola) [inline core/src/test/java/org/camunda/spin/test/script/JavascriptScriptRuleNashornTest.java:24]: Let's remove the author comments.
```suggestion
```

File: dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeCreateJavascriptTest.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ */
+package org.camunda.spin.javascript.nashorn.json.tree;
+
+import org.camunda.spin.json.tree.JsonTreeCreateScriptTest;
+import org.camunda.spin.test.ScriptEngine;
+
+/**
+ * @author Thorben Lindhauer
+ *
+ */
```
Comment (koevskinikola) [inline dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeCreateJavascriptTest.java:25]: Let's remove the author comments for new files.
```suggestion
```

File: dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeEditListPropertyJavascriptTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+import org.camunda.spin.json.tree.JsonTreeEditListPropertyScriptTest;
+import org.camunda.spin.test.ScriptEngine;
+import org.junit.Ignore;
+import org.junit.Test;
+
+/**
+ * @author Stefan Hentschel
+ *
+ */
```
Comment (koevskinikola) [inline dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeEditListPropertyJavascriptTest.java:27]: Same as above:
```suggestion
```

File: dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeMapJsonToJavaJavascriptTest.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+package org.camunda.spin.javascript.nashorn.json.tree;
+
+import org.camunda.spin.json.tree.JsonTreeMapJsonToJavaScriptTest;
+import org.camunda.spin.test.ScriptEngine;
+
+/**
+ * @author Sebastian Menski
+ */
```
Comment (koevskinikola) [inline dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeMapJsonToJavaJavascriptTest.java:24]:
```suggestion
```

File: dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeMapObjectToJsonJavascriptTest.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+import org.camunda.spin.json.tree.JsonTreeMapObjectToJsonScriptTest;
+import org.camunda.spin.test.ScriptEngine;
+import org.junit.Ignore;
+
+/**
+ * Note: Jackson is apparently not able to map Nashorn's internal representations
+ * of javascript objects and arrays.
+ *
+ * @author Thorben Lindhauer
```
Comment (koevskinikola) [inline dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeMapObjectToJsonJavascriptTest.java:27]:
```suggestion
```

File: dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeReadPropertyJavascriptTest.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ */
+package org.camunda.spin.javascript.nashorn.json.tree;
+
+import org.camunda.spin.json.tree.JsonTreeReadPropertyScriptTest;
+import org.camunda.spin.test.ScriptEngine;
+
+/**
+ * @author Thorben Lindhauer
+ *
+ */
```
Comment (koevskinikola) [inline dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeReadPropertyJavascriptTest.java:25]:
```suggestion
```

File: dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeRemovePropertyJavascriptTest.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ */
+package org.camunda.spin.javascript.nashorn.json.tree;
+
+import org.camunda.spin.json.tree.JsonTreeRemovePropertyScriptTest;
+import org.camunda.spin.test.ScriptEngine;
+
+/**
+ * @author Thorben Lindhauer
+ *
+ */
```
Comment (koevskinikola) [inline dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeRemovePropertyJavascriptTest.java:25]:
```suggestion
```

File: dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeSetPropertyJavascriptTest.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ */
+package org.camunda.spin.javascript.nashorn.json.tree;
+
+import org.camunda.spin.json.tree.JsonTreeSetPropertyScriptTest;
+import org.camunda.spin.test.ScriptEngine;
+
+/**
+ * @author Thorben Lindhauer
+ *
+ */
```
Comment (koevskinikola) [inline dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeSetPropertyJavascriptTest.java:25]:
```suggestion
```

File: dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/json/tree/JsonTreeMapObjectToJsonJavascriptTest.java [hunk 9]
```diff
@@ -16,18 +16,18 @@
...
 package org.camunda.spin.javascript.json.tree;
 
-import org.camunda.spin.impl.test.ScriptEngine;
 import org.camunda.spin.json.tree.JsonTreeMapObjectToJsonScriptTest;
+import org.camunda.spin.test.ScriptEngine;
 import org.junit.Ignore;
 
 /**
- * Note: Jackson is apparently not able to map Nashorn's internal representations
+ * Note: Jackson is apparently not able to map graal.js's internal representations
```
Comment (koevskinikola) [inline dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/json/tree/JsonTreeMapObjectToJsonJavascriptTest.java:24]: The comment mentions `graal.js` but the `@ScriptEngine` annotation receives `nashorn` as an input argument. Is this correct? It's a bit confusing.

File: dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeJsonPathJavascriptTest.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+ */
+package org.camunda.spin.javascript.nashorn.json.tree;
+
+import org.camunda.spin.json.tree.JsonTreeJsonPathScriptTest;
+import org.camunda.spin.test.ScriptEngine;
+
+/**
+ * @author Thorben Lindhauer
+ *
+ */
```
Comment (koevskinikola) [inline dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/nashorn/json/tree/JsonTreeJsonPathJavascriptTest.java:25]:
```suggestion
```

**Review by koevskinikola — CHANGES_REQUESTED** [review]
The Spin `org.camunda.spin.test` package should be reverted to `org.camunda.spin.impl.test`. Users might assume that this is now a public API. Especially since now, we include a test-jar in the BOM.

I also added code suggestions to remove the author comments in new files.

File: dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/json/tree/JsonTreeMapObjectToJsonJavascriptTest.java (hunk 9 above)
Comment (tmetzke) [reply dataformat-json-jackson/src/test/java/org/camunda/spin/javascript/json/tree/JsonTreeMapObjectToJsonJavascriptTest.java:24]: It's supposed to run on Graal.js

**Review by koevskinikola — APPROVED** [review]
All good!

**Pull request merged**